build/commands/lib/test.js

```diff
@@ -212,89 +215,100 @@ const runTests = (passthroughArgs, suite, buildConfig, options) => {
+        fs.appendFileSync(allResultsFilePath, `${testSuite}.xml\n`)
       }
+
       if (config.targetOS === 'android' && !isJunitTestSuite) {
         assert(
           config.targetArch === 'x86'
             || config.targetArch === 'x64'
             || options.manual_android_test_device,
           'Only x86 and x64 builds can be run automatically. For other builds please run test device manually and specify manual_android_test_device flag.',
         )
+
+        if (!options.manual_android_test_device) {
+          runArgs.push(
+            `--avd-config=tools/android/avd/proto/${options.android_test_emulator_name}.textpb`,
+          )
+        }
       }
-      if (
-        config.targetOS === 'android'
-        && !isJunitTestSuite
-        && !options.manual_android_test_device
-      ) {
-        // Specify emulator to run tests on
-        braveArgs.push(
-          `--avd-config=tools/android/avd/proto/${options.android_test_emulator_name}.textpb`,
-        )
-      }
-      let runOptions = config.defaultOptions
+
       if (config.isTeamcity) {
         // Stdout and stderr must be separate for a test launcher.
         runOptions.stdio = 'inherit'
       }
-      if (options.output_xml) {
-        // When test results are saved to a file, callers (such as CI) generate
-        // and analyze test reports as a next step. These callers are typically
-        // not interested in the exit code of running the tests, because they
-        // get the information about test success or failure from the output
-        // file. On the other hand, callers are interested in errors that
-        // produce an exit code, such as test compilation failures. By ignoring
-        // the test exit code here, we give callers a chance to distinguish test
-        // failures (by looking at the output file) from compilation errors.
-        runOptions.continueOnFail = true
-      }
-      let prog = util.run(
-        path.join(config.outputDir, getTestBinary(testSuite)),
-        braveArgs,
-        runOptions,
-      )
-      if (options.output_xml) {
-        // Add filename of xml output of each test suite into the results file
-        fs.appendFileSync(allResultsFilePath, `${testSuite}.xml\n`)
+
+      let prog = util.run(getTestBinary(testSuite), runArgs, runOptions)
+
+      // convert json results to xml
+      if (prog.status === 0 && convertJSONToXML) {
+        prog = util.run('vpython3', [path.join('script', 'json2xunit.py')], {
+          ...config.defaultOptions,
+          cwd: config.braveCoreDir,
+          stdio: [
+            fs.openSync(`${outputFilename}.json`, 'r'),
+            fs.openSync(`${outputFilename}.xml`, 'w'),
+            'inherit',
+          ],
+        })
+      } else {
+        testsDidFail = prog.status !== 0
       }
-      // Don't run other tests if one has failed already.
-      return prog.status === 0
     })
+    return testsDidFail ? -1 : 0
```

Comment:
I am not sure if returning a failure here would prevent CI from continuing onto parsing the results. cc: @wknapik

Reply:
Previously, the process would return success regardless of test failures. Now it will return failure if a test fails, which will make CI mark it red and prevent it from parsing results.

Reply:
https://github.com/brave/brave-core/pull/29870/files#diff-ae08efb1b3db917a9b74c8758cdfc636bc34464ebf56fcc454d4acc619e3d05bL296

Reply:
@mkarolin what I'm wondering is if we can differentiate between a non-zero output from `npm run test` from a normal test failure vs a failure for some other reason. I'm building right now to check`